psyneulink/core/llvm/helpers.py

```diff
@@ -239,6 +228,42 @@ def is_boolean(x):
         type_t = x.type.pointee
     return isinstance(type_t, ir.IntType) and type_t.width == 1
 
+def get_array_shape(x):
+    x_ty = x.type
+    if is_pointer(x):
+        x_ty = x_ty.pointee
+
+    assert isinstance(x_ty, ir.ArrayType), f"Tried to get shape of non-array type: {x_ty}"
+    dimensions = []
+    while hasattr(x_ty, "count"):
+        dimensions.append(x_ty.count)
+        x_ty = x_ty.element
+
+    return tuple(dimensions)
+
+def array_from_shape(shape, element_ty):
+    array_ty = element_ty
+    for dim in reversed(shape):
+        array_ty = ir.ArrayType(array_ty, dim)
+    return array_ty
+
+def recursive_iterate_arrays(ctx, builder, u, *args):
+    """Recursively iterates over all elements in scalar arrays of the same shape"""
+    assert isinstance(u.type.pointee, ir.ArrayType), "Can only iterate over arrays!"
+    assert all(len(u.type.pointee) == len(v.type.pointee) for v in args), "Tried to iterate over differing types!"
+    with array_ptr_loop(builder, u, str(u) + "," + str(args) + "_recursive_zip") as (b, idx):
+        u_ptr = b.gep(u, [ctx.int32_ty(0), idx])
+        arg_ptrs = (b.gep(v, [ctx.int32_ty(0), idx]) for v in args)
+        if is_scalar(u_ptr):
+            yield (u_ptr, *arg_ptrs)
+        else:
+            yield from recursive_iterate_arrays(ctx, b, u_ptr, *arg_ptrs)
+
+def call_elementwise_operation(ctx, builder, x, operation, output_ptr):
```

Comment:
This can be marked for removal. It doesn't really add anything on top of the iterator and leads to overly long lines. At that point splitting the invocation into two lines is more readable anyway.
But that is for later cleanup (not in this PR)